File: client/src/game/entities/backgrounditem.ts
import * as PIXI from "pixi.js";
import { EntityType } from "../../network/game/EntityType";
import { BackgroundItemType, Direction, DrawLayer, Entity } from "../entity";
import { getTexture } from "../resources";

export interface TowerProps {
  x?: number;
  y?: number;
  type?: number;
  direction?: number;
}

export class BackgroundItem extends Entity {
  x = 0;
  y = 0;
  direction = Direction.LEFT;
  type = EntityType.BACKGROUND_ITEM;
  subType = BackgroundItemType.NORMAL_TOWER;

  container = new PIXI.Container();
  sprite = new PIXI.Sprite(getTexture("controlTower.gif"));

  constructor() {
    super();

    this.container.addChild(this.sprite);
    this.container.zIndex = DrawLayer.ControlTower;
  }

  getContainer(): PIXI.Container {
    return this.container;
  }

  redraw() {

    switch (this.subType) {
      case BackgroundItemType.DESERT_TOWER:
        this.sprite.texture = getTexture("controlTowerDesert.gif");
        break;
      case BackgroundItemType.PALM_TREE:
        this.sprite.texture = getTexture("palmtree.gif");
      default: {
      }
    }

    const xPos = this.x - Math.round(this.sprite.width / 2);
    const yPos = this.y - this.sprite.height;
    this.sprite.position.set(xPos, yPos);

    if (this.direction == Direction.RIGHT) {
      this.sprite.anchor.x = 1;
      this.sprite.scale.x *= -1;
    }
  }

  destroy() { }
}

File: game/src/main/java/com/lentokonepeli/entities/BackgroundItem.java
package com.lentokonepeli.entities;

import com.lentokonepeli.BackgroundItemType;
import com.lentokonepeli.Direction;
import com.lentokonepeli.Entity;
import com.lentokonepeli.EntityType;
import com.lentokonepeli.network.NetProp;
import com.lentokonepeli.network.NetType;

public class BackgroundItem extends Entity {

    private NetProp<Integer> x = new NetProp<>("x", NetType.i16);
    private NetProp<Integer> y = new NetProp<>("y", NetType.i16);
    private NetProp<Integer> subType = new NetProp<>("subType", NetType.u8);
    private NetProp<Integer> direction = new NetProp<>("direction", NetType.u8);

    public BackgroundItem(int x, int y, BackgroundItemType type, Direction direction) {
        super(EntityType.BACKGROUND_ITEM);

        props.add(this.x);
        props.add(this.y);
        props.add(this.subType);
        props.add(this.direction);

        this.x.set(x);
        this.y.set(y);
        this.subType.set(type.ordinal());
        this.direction.set(direction.ordinal());
    }
}
